### ISSUE TITLE
Range indexing ideas

### ISSUE THREAD

**Issue by kubouch** [nushell/nushell#3378]
Idea and context from this comment: https://github.com/nushell/nushell/pull/3377#discussion_r624745521

Range would be reversed when lower bound > upper bound. For example `seq 10 | range 5..2` would yield 5, 4, 3, 2.

Also, maybe we could introduce Python-style negative indexing? For example `seq 10 | range ..-2` would be equal to `seq 10 | range ..8`.

The `nu_protocol::Range` implementation does not prevent "from" > "to" already, so I think it's just about enabling this in individual commands.

**Comment by alexshadley** [comment]
One potential obstacle to negative indexing would be parsing, e.g. with `seq 1 10 | range -3..` nushell interprets `-3` as a flag.

Do we have reasonable workarounds for this?

**Comment by ilius** [comment]
For seq (or when the argument is just a negative int) we can use int command like this
https://github.com/ilius/nu-addons/blob/main/functions.nu#L109

But in general, I think it's best to implement `--` which is standard in Unix, for example `seq 1 10 | range -- -3..`.
Anything after -- should be considered non-flag.

**Comment by kubouch** [comment]
Right now, the workaround is `(-3)..` until that PR lands.

**Comment by alexshadley** [comment]
In tinkering around with this, I realized that negative indexing requires us to collect our `InputStream` into a `Vec` (or some sort of collection). How bad is this, performance-wise? We could at least avoid loading the entire stream into a vec when we don't have any negative indices if we wanted to mitigate performance issues.

**Comment by kubouch** [comment]
@alexshadley Yes, negative indexing requires collecting the entire stream (you need to locate the end of the stream). I think we should avoid this as much as possible, so with non-negative indices it should proceed as usual (i.e., without collecting). I don't have much knowledge about the internals, but imagine `seq 1000000 | range 2..5`, with collecting, you would have to wait till the end of the stream, instead of breaking the stream after the 5th sample without collecting.

The negative indexing could also be used in other commands, such as `nth`, `first`, etc. Basically all of the "selecting" types of commands could support it.

**Comment by kubouch** [comment]
We have negative indexing, but still need (?) support for from > to reversing the sequence.

**Comment by jntrnr** [comment]
Keeping this open, as we're still not able to really use range indicies.

**Comment by CAESIUS-TIM** [comment]
4dcaf2a201fafe274904e1972e9148d981697a8c

Please rename the issue title, since `range` has been renamed to `slice`.
